### ISSUE TITLE
Virtual Camera will start one time and will not restart until reboot

### ISSUE THREAD

**Issue by bmn001** [obsproject/obs-studio#4808]
### Operating System Info

Ubuntu 21.04

### OBS Studio Version

27.0.0

### OBS Studio Log URL

https://obsproject.com/logs/Cy96Hu3XHeD7c-Zb

### Expected Behavior

I should be able to toggle Virtual Camera on and off.

### Current Behavior

I'm able to start and stop Virtual Camera one time.  After that, it will not start again until my next reboot.  Each click on the "Start Virtual Camera" button inserts a single line of text into the log:
`04:00:54 PM.094: Failed to start virtual camera`

### Steps to Reproduce

1. Reboot
2. Launch OBS
3. Click "Start Virtual Camera"
4. Click "Stop Virtual Camera"
5. Click "Start Virtual Camera".  The button text remains the same ("Start Virtual Camera") and the VC is never started.

### Anything else we should know?

This began happening immediately after an OS upgrade from Ubuntu 20.10 to 21.04.  Things worked fine in Ubuntu 20.10 using the same version of OBS (27.0.0).

The attached log shows me launching OBS immediately after a reboot.
I click to start the virtual camera.
I enter my sudo pw.
I click to stop the virtual camera.
So far so good.
I click to start the virtual camera - three times - and all three times result in failure and an error message in the log: "Failed to start virtual camera."

The output of `v4l2-ctl --list-devices` immediately after reboot:
```
UVC Camera (046d:0821) (usb-0000:00:14.0-8.4):
	/dev/video4
	/dev/video5

Integrated_Webcam_HD: Integrate (usb-0000:00:14.0-9):
	/dev/video0
	/dev/video1
	/dev/video2
	/dev/video3
```

The output of `v4l2-ctl --list-devices` after I've tried and failed to get the Virtual Camera to start:
```
OBS Virtual Camera (platform:v4l2loopback-000):
    /dev/video6

UVC Camera (046d:0821) (usb-0000:00:14.0-8.4):
    /dev/video4
    /dev/video5

Integrated_Webcam_HD: Integrate (usb-0000:00:14.0-9):
    /dev/video0
    /dev/video1
    /dev/video2
    /dev/video3
```

This is on a Dell XPS 13 laptop, model 9300.  If it matters, I'm not using the integrated webcam, I'm using Logitech C910 USB cam.

**Comment by bmn001** [comment]
If I quit OBS, then remove the modprobe rule with:
``sudo modprobe -r v4l2loopback``
I'm able to restart OBS and launch the Virtual Camera successfully (without rebooting) - but again, only one time.

**Comment by troyhitch** [comment]
Hi, for what it's worth, I'm experiencing the exact same thing having updated to 21.04 earlier today. I have been able to reliably restart the VC after a reboot. And I've been able to successfully restart it after killing OBS and running `sudo modprobe -r v4l2loopback` per above and have been able to do it more than once.

**Comment by bmn001** [comment]
OBS 27.0.1 came out today and the issue is still there.

**Comment by jbrains** [comment]
I have the same problem on Pop!_OS 21.04 and the workaround of `sudo modprobe -r v4l2loopback` behaves for me as the others have described.

I built from source just now (commit https://github.com/obsproject/obs-studio/commit/16a49eb10d4609198fc6b1502319d5e73ce0c3ee) and the problem persists.

**Comment by WebrangerC** [comment]
Same issue with 21.04 and OBS 27.0.1

**Comment by pmichaud** [comment]
On my system, the problem appears to be related to the "exclusive_caps=1" option of the v4l2loopback module.

If I remove "exclusive_caps=1" from whatever is creating the virtual camera device, then obs is able to start / stop / restart the virtual camera with no difficulty, both within a single obs-studio invocation and across multiple invocations.

Unfortunately, removing "exclusive_caps=1" means that the v4l2loopback output device is no longer accessible by Chrome and some other programs.  The version of Zoom that I have still works -- it finds the virtual device just fine -- but Chrome no longer finds it.

------

If I understand correctly, exclusive_caps=1 tells v4loopback to start the virtual device in "CAPTURE" capabilities only, then switch to "OUTPUT only" capabilities once something (like obs) starts sending data to the virtual device.  However, when the data stream stops, the virtual device remains in "OUTPUT only" mode and I think this prevents OBS from either finding or re-opening the device on second and subsequent attempts.  The only reliable solution is to restart the virtual device altogether, either via a reboot or a rmmod/modprobe of the device as mentioned by others above.

Ideally I think the virtual device should reset itself to advertising CAPTURE capabilities when it no longer has an active input data stream -- but that would be a fix to v4l2loopback instead of obs.   Alternatively, obs should find/open the device for streaming even if it's not currently advertising CAPTURE capabilities.

Hope this helps point the way to a solution.

Pm

**Comment by pmichaud** [comment]
To follow-up a bit on my previous post:  If I set up the loopback device without "exclusive_caps=1", then "v4l2-ctl -d /dev/video0 --all" gives me the following output:

```
pmichaud@kiwi:~$ sudo v4l2-ctl -d /dev/video0 --all
Driver Info:
	Driver name      : v4l2 loopback
	Card type        : Dummy video device (0x0000)
	Bus info         : platform:v4l2loopback-000
	Driver version   : 5.11.22
	Capabilities     : 0x85200003
		Video Capture
		Video Output
		Read/Write
		Streaming
		Extended Pix Format
		Device Capabilities
	Device Caps      : 0x05200003
		Video Capture
		Video Output
		Read/Write
		Streaming
		Extended Pix Format
Priority: 2
```

This part of the output remains the same no matter how many times "Start Virtual Camera" and "Stop Virtual Camera" are selected.  Zoom recognizes the virtual loopback device just fine, but Chrome does not find it.

If I add "exclusive_caps=1" when loading the virtual device via modprobe, the device initially appears as follows:

```
pmichaud@kiwi:~$ sudo v4l2-ctl -d /dev/video0 --all
Driver Info:
	Driver name      : v4l2 loopback
	Card type        : Dummy video device (0x0000)
	Bus info         : platform:v4l2loopback-000
	Driver version   : 5.11.22
	Capabilities     : 0x85200002
		Video Output
		Read/Write
		Streaming
		Extended Pix Format
		Device Capabilities
	Device Caps      : 0x05200002
		Video Output
		Read/Write
		Streaming
		Extended Pix Format
Priority: 2
```

Once I hit "Start Virtual Camera" in OBS, the device settings change to:

```
pmichaud@kiwi:~$ sudo v4l2-ctl -d /dev/video0 --all
Driver Info:
	Driver name      : v4l2 loopback
	Card type        : Dummy video device (0x0000)
	Bus info         : platform:v4l2loopback-000
	Driver version   : 5.11.22
	Capabilities     : 0x85200001
		Video Capture
		Read/Write
		Streaming
		Extended Pix Format
		Device Capabilities
	Device Caps      : 0x05200001
		Video Capture
		Read/Write
		Streaming
		Extended Pix Format
Priority: 2
```

Stopping the Virtual Camera in OBS (or exiting OBS) results in:

```
pmichaud@kiwi:~$ sudo v4l2-ctl -d /dev/video0 --all
Driver Info:
	Driver name      : v4l2 loopback
	Card type        : Dummy video device (0x0000)
	Bus info         : platform:v4l2loopback-000
	Driver version   : 5.11.22
	Capabilities     : 0x85200000
		Read/Write
		Streaming
		Extended Pix Format
		Device Capabilities
	Device Caps      : 0x05200000
		Read/Write
		Streaming
		Extended Pix Format
Priority: 2
```

Note that the video device no longer advertises either output or capture capabilities.  As a result, OBS won't or can't re-open the device to restart the loopback video stream.

So this looks to me like something that needs to be fixed within v4l2loopback. 

Pm

**Comment by WebrangerC** [comment]
@pmichaud THANK YOU. 
I haven't duplicated/verified the exclusive_caps BUT using modprobe to re-initialize: 
  - sudo modprobe -r v4l2loopback
  - sudo modprobe v4l2loopback
works well for Zoom, even if Zoom has been started and I wanted to switch to using OBS.

**Comment by prasket** [comment]
I was having this issue as well using POP_OS and 27.1.0 but the modprobe was my hotfix. I just one of my machine using Debian Stable and I just built OBS 27.1.1 and the start/stop of the virtual camera appears to be working just fine.

**Comment by bmn001** [comment]
Just tested OBS 27.1.3 and it looks like the issue is still there.

**Comment by cmcqueen** [comment]
> Just tested OBS 27.1.3 and it looks like the issue is still there.

It looks to be a bug in the `v4l2loopback` kernel driver. The issue isn't really in OBS, of any version.

**Comment by bmn001** [comment]
Right you are. Looks like a v4l2loopback bug referencing this one has been filed. https://github.com/umlaeute/v4l2loopback/issues/442

**Comment by eduncan911** [comment]
Subscribed, same issue w/Pop_OS 21.04 (aka Ubuntu 21.04).  The `modprobe` command works to restart it.

**Comment by UnderColonel** [comment]
Thanks for this! From Russia with love) !
Ubuntu 21.10
5.15.5-051505-generic
OBS Studio 27.1.3
version 0.12.5-1 works as expected; 
version 0.12.5.-1ubuntu3 does not work as expected;

**Comment by koo5** [comment]
> v4l2loopback-dkms_0.12.5-1_all
can you explain what you meant with the second version? on my system i have ```
Package: v4l2loopback-dkms
Version: 0.12.5-1ubuntu3```

**Comment by UnderColonel** [comment]
> > v4l2loopback-dkms_0.12.5-1_all
> > can you explain what you meant with the second version? on my system i have ` Package: v4l2loopback-dkms Version: 0.12.5-1ubuntu3`
I have my system version 0.12.5-1ubuntu3 default too,
but, work only: ftp.ubuntu.com/ubuntu/ubuntu/pool/universe/v/v4l2loopback/v4l2loopback-dkms_0.12.5-1_all.deb
link: "http://ftp.ubuntu.com/ubuntu/ubuntu/pool/universe/v/v4l2loopback/v4l2loopback-dkms_0.12.5-1ubuntu3_all.deb" is my typo (

**Comment by joserobjr** [comment]
> @bmn001 Thanks for this! I have confirmed the following:
> 
>     * version `0.12.5-1` works as expected; it does _not_ exhibit this issue
> 
>     * version `0.12.5.-1ubuntu1` does _not_ work as expected; it _does_ exhibit this issue
> 
> 
> I did the following, after downloading version `0.12.5-1` of the package from https://packages.ubuntu.com/groovy/all/v4l2loopback-dkms/download:
> 
> ```
> $ sudo modprobe -r v4l2loopback
> $ sudo dpkg -i $HOME/Downloads/v4l2loopback-dkms_0.12.5-1_all.deb
> $ sudo apt-mark hold v4l2loopback-dkms
> $ sudo apt-get update
> $ sudo apt-get upgrade
> Reading package lists... Done
> Building dependency tree... Done
> Reading state information... Done
> Calculating upgrade... Done
> The following packages have been kept back:
>   v4l2loopback-dkms
> 0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
> ```
> 
> That seems to work for me. I can switch the Virtual Camera off then on repeatedly in OBS Studio.

I could not find `0.12.5-1` there, the link was not opening for me, but I found `0.12.3-1ubuntu0.4` which is marked as a security update, I tried this one (downgrading from `0.12.5.-1ubuntu1` and it's working for me. I'm using Kubuntu 21.04

![image](https://user-images.githubusercontent.com/3679022/148450181-24b3d7ae-628e-4127-ba10-bd229f54150b.png)

**Comment by jbrains** [comment]
I found a Debian package for version 0.12.5-1 at https://packages.debian.org/bullseye/all/v4l2loopback-dkms/download

**Comment by RytoEX** [comment]
After reviewing this, it seems that comments on [Ubuntu v4l2loopback Bug 1946660](https://bugs.launchpad.net/ubuntu/+source/v4l2loopback/+bug/1946660/) and https://github.com/umlaeute/v4l2loopback/issues/404 (specifically, https://github.com/umlaeute/v4l2loopback/issues/404#issuecomment-1001995043 and the following two comments) conclude that this is a bug in https://github.com/umlaeute/v4l2loopback/.  It looks like this is meant to be addressed in https://github.com/umlaeute/v4l2loopback/pull/359 which is still a WIP/RFC in Draft status.  I'm not sure if there's anything we can do about this on OBS' side.

**Comment by bmn001** [comment]
For Ubuntu users coming here after a recent upgrade to 22.04 LTS who are now running into this issue, the temporary downgrade/workaround discussed above still works:

* Grab `v4l2loopback-dkms_0.12.5-1_all.deb` from https://packages.debian.org/sid/all/v4l2loopback-dkms/download
* Install it: `sudo dpkg -i v4l2loopback-dkms_0.12.5-1_all.deb`
* Hold it back so system updates don't wipe it out: `sudo apt-mark hold v4l2loopback-dkms`
* Remove the hold in the future when upstream gets fixed: `sudo apt-mark unhold v4l2loopback-dkms`

The last version of v4l2loopback that worked properly is 0.12.5-1 and that's what we're installing above.
(Ubuntu 21.04 includes `0.12.5-1ubuntu1` and Ubuntu 22.04 includes `0.12.5-1ubuntu5`)

**Comment by Omega7379** [comment]
> For Ubuntu users coming here after a recent upgrade to 22.04 LTS who are now running into this issue, the temporary downgrade/workaround discussed above still works:
> 
>     * Grab `v4l2loopback-dkms_0.12.5-1_all.deb` from https://packages.debian.org/sid/all/v4l2loopback-dkms/download
> 
>     * Install it: `sudo dpkg -i v4l2loopback-dkms_0.12.5-1_all.deb`
> 
>     * Hold it back so system updates don't wipe it out: `sudo apt-mark hold v4l2loopback-dkms`
> 
>     * Remove the hold in the future when upstream gets fixed: `sudo apt-mark unhold v4l2loopback-dkms`
> 
> 
> The last version of v4l2loopback that worked properly is 0.12.5-1 and that's what we're installing above. (Ubuntu 21.04 includes `0.12.5-1ubuntu1` and Ubuntu 22.04 includes `0.12.5-1ubuntu5`)

for anyone looking to see if this works with Pop_OS, the error also occurs there but can be fixed with the above steps. Great work everyone on this thread to get a temporary and a permanent fix dealt with.

**Comment by jbrains** [comment]
I have just upgraded to Pop!_OS 22.04, I have reinstalled `v4l2loopback-dkms` version `0.12.5-1`, and I have rebooted.

OBS Studio will not restart the Virtual Camera unless I unload the kernel module by handle and restart OBS Studio.

Am I missing something? Did something change? What can I try? I don't remember all the details from last year, when I was actively trying to figure this out.

**Comment by Lovinity** [comment]
On Ubuntu Studio 22.04, can confirm this issue is present. Unfortunately, both workarounds presented do not sufficiently work:

`sudo modprobe -r v4l2loopback` throws "modprobe: FATAL: Module v4l2loopback is in use."

And, while rolling back to the Debian package fixed this issue, OBS cannot access my camera / video capture device with this package.

**Comment by RedKrieg** [comment]
@flexiondotorg I've just tried your PPA and while it does let me start virtual camera in OBS multiple times without having to rmmod v4l2loopback, it also prevents any other application from seeing the virtual camera device for me.

**Comment by flexiondotorg** [comment]
@RedKrieg Bother. You're right. I can't access the virtual camera 😟

**Comment by flexiondotorg** [comment]
@RedKrieg Updated versions of `v4l2loopback` currently building in the PPA that add a backport of https://github.com/umlaeute/v4l2loopback/pull/485

I can start/stop the Virtual Cam multiple times in OBS Studio and when the Virtual Cam is running I can access it in several video conference apps I use.

**Comment by RedKrieg** [comment]
@flexiondotorg I just tried your PPA again and everything works as expected!  Thank you for your work here.

**Comment by flexiondotorg** [comment]
@kkartaltepe Looks like this one can be closed as the bug is in `v4l2loopback` and not OBS.

**Comment by RytoEX** [comment]
Closing per the above comments.

**Comment by kkartaltepe** [comment]
I think we were leaving these and other ones like Qt bugs open while we wait for the fixes to make it to upstream distros. Since this is still a bug in ubuntu 22 I thought we were leaving it open.

**Comment by eduncan911** [comment]
Maybe add a tag, to triage it later?

**Comment by Fenrirthviti** [comment]
I'm fine leaving this open as a tracking point, so we don't have new issues opened.

**Comment by WolfdaleA3** [comment]
I would also like to add that this bug also occurs in 22.04 Jammy.

**Comment by hubyhuby** [comment]
It  kinda work with this workaround under Ubuntu 22.04 (I had no problem on ubunt 20.04)

1) Type :  sudo modprobe -r v4l2loopback
2) Then launch OBS (in same terminal) 
3) Start camera : it asked me again modprob authorization (enter passworrd)
4) Start the "virtual camera"  => It works
5) Stop the "virtual camera" => you cannot start it again
6) shutdown OBS and go back to 1 to start again the camera 

![image](https://user-images.githubusercontent.com/26340942/180976261-cab32dc6-095a-479c-8256-1f76fe65f99b.png)

**Comment by jbrains** [comment]
Pop!_OS 22.04 appears to have incorporated this fix into their standard package. I have reverted the PPA mentioned in https://github.com/obsproject/obs-studio/issues/4808#issuecomment-1184588057 and version `0.12.5-1ubuntu7pop1~1659015816~22.04~951a0b7` of the package appears to behave as expected, at least on my machine.

```
$ sudo ppa-purge -o flexiondotorg -s ppa.launchpadcontent.net -p v4l2loopback
```

**Comment by Eschguy** [comment]
This is an issue with Fedora 36 as well

**Comment by jbrains** [comment]
Now on Pop!_OS 22.04, the problem has returned. I didn't notice when, so I don't know what caused the problem.

I can confirm that the PPA version `0.12.5-1ubuntu7.1~jammy1.1` continues to work, but the standard Pop!_OS 22.04 version `0.12.5-1ubuntu7pop1~1659015816~22.04~951a0b7` fails (exhibits this issue).

If you're on Pop!_OS 22.04, then install the PPA version of this package and hold it.

**Comment by RytoEX** [comment]
> Now on Pop!_OS 22.04, the problem has returned. I didn't notice when, so I don't know what caused the problem.
> 
> I can confirm that the PPA version `0.12.5-1ubuntu7.1~jammy1.1` continues to work, but the standard Pop!_OS 22.04 version `0.12.5-1ubuntu7pop1~1659015816~22.04~951a0b7` fails (exhibits this issue).
> 
> If you're on Pop!_OS 22.04, then install the PPA version of this package and hold it.

Please report this to the package maintainer.

**Comment by jbrains** [comment]
> > Now on Pop!_OS 22.04, the problem has returned. I didn't notice when, so I don't know what caused the problem.
> > I can confirm that the PPA version `0.12.5-1ubuntu7.1~jammy1.1` continues to work, but the standard Pop!_OS 22.04 version `0.12.5-1ubuntu7pop1~1659015816~22.04~951a0b7` fails (exhibits this issue).
> > If you're on Pop!_OS 22.04, then install the PPA version of this package and hold it.
> 
> Please report this to the package maintainer.

@RytoEX Pull request https://github.com/umlaeute/v4l2loopback/pull/485 is already open; the PPA version of the package already incorporates that fix; I'm not sure what else to report to the package maintainer. I leave comments here only because people searching are more likely to find this issue than https://github.com/umlaeute/v4l2loopback/issues/484.

**Comment by kkartaltepe** [comment]
> the PPA version of the package already incorporates that fix; I'm not sure what else to report to the package maintainer

You report to the Ubuntu maintainers that they should incorporate the fix from upstream just like the PPA.

**Comment by jbrains** [comment]
> Then you kindly ask the v4l2 upstream to create a release which incorporates the fixes from main.

Haha! Yes, indeed. I somehow failed to notice that that pull request was actually merged; I thought it was still stuck waiting for feedback.

**Comment by zDEFz** [comment]
Mine would not let me run OBS as sudo / root and then I could no more start the virtual camera.
Workaround? sudo modprobe v4l2loopback video_nr=2 card_label="OBS Virtual Camera"

**Comment by kkartaltepe** [comment]
As mentioned elsewhere you must ask your package manager to _include the fixes_ and _upstream to make a release with the fixes_, Asking your package manager to update to the latest version is not sufficient because upstream has not made a release with the fixes. Do not look at when upstream last release and assume it includes things on main.

**Comment by jbrains** [comment]
@kkartaltepe I think I did both those things. Are you assuming that I didn't or did you read something that tells you that I didn't?

**Comment by kkartaltepe** [comment]
I don't think your partner interpreted it that way in the discussion linked, which is why I specified.

**Comment by zDEFz** [comment]
If theres a fix, will this also be pushed in the AUR?

**Comment by BenBE** [comment]
As I'm the author behind the fixes in https://github.com/umlaeute/v4l2loopback/pull/477 and https://github.com/umlaeute/v4l2loopback/pull/485, I can provide the following additional information, that might help as workaround even on systems where the above two patches aren't implemented yet:

A careful reading of the [kernel documentation for V4L2](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-streamon.html) notes in section 7.54.4 that the `VIDIOC_STREAMON` ioctl must be called by both the output and the capture application before a stream is generated:
> Capture hardware is disabled and no input buffers are filled (if there are any empty buffers in the incoming queue) until VIDIOC_STREAMON has been called. Output hardware is disabled and no video signal is produced until VIDIOC_STREAMON has been called.
>
> Memory-to-memory devices will not start until VIDIOC_STREAMON has been called for both the capture and output stream types.

Comparing this with the implementation of `v4l2loopback` shows that IF you perform this ioctl **before** any write operation, the internal state is properly updated to mark this opener as a stream producer and thus the bug causing the loss of output capability (when closing the device fd) is avoided. What the two patches do is (477) force the `VIDIOC_STREAMON` ioctl to appear before any `write` can be performed and (495) work around the issue that most applications in user land seem to disregard the kernel documentation and thus make the `VIDIOC_STREAMON` ioctl call implicit on first `write`.

Thus what OBS can do on broken systems is simply perform the `VIDIOC_STREAMON` ioctl before starting to stream to the output device and things should work even without the updated `v4l2loopback` kernel module.

I hope this information helps with fixing the root cause on the OBS side (even though the driver now works around this quirk from userland applications).

**Comment by kkartaltepe** [comment]
>  (even though the driver now works around this quirk from userland applications).

I appreciate you fixing the regression in v4l2loopback, but as explained this is false. There is no such release of v4l2 with your changes which is why people who want your fixes should request upstream to make such a release.

**Comment by BenBE** [comment]
> > (even though the driver now works around this quirk from userland applications).
> 
> I appreciate you fixing the regression in v4l2loopback, but as explained this is false. There is no such release of v4l2loopback with your changes which is why people who want your fixes should request upstream to make such a release. Or to patch in your fixes into their packages.

I never claimed that there already was a release, just that the driver (using the latest development code) changed the behavior.

Also my note was additionally that aside from fixing v4l2loopback, you also just can include a change in OBS to ensure the mentioned ioctl is executed before writing any image data and the problem will be gone as well. Doing so would even be compliant to the kernel documentation (which I linked for reference).

But I'm not upstream for v4l2loopback; I just review its code (together with some others) and we found a few more issues besides this one.

**Comment by kkartaltepe** [comment]
Yes we can fix OBS and leave everyone else who uses v4l2loopback broken, thats always been possible. We already have a PR in flight for applying your recommendations.

The point was always to fix the problem at its source because I don't want to deal with users who have used other v4l2loopback tools in the meantime that broke their system. For example our users often use droidcam with OBS, droidcam does not have your suggested patch so when users bust their v4l2loopback devices with droidcam how are we supported work around that?

**Comment by jbrains** [comment]
> I don't think your partner interpreted it that way in the discussion linked, which is why I specified.

Thanks for clarifying. It looks like I got what I was hoping for: people are talking and things are moving.

**Comment by BenBE** [comment]
> For example our users often use droidcam with OBS, droidcam does not have your suggested patch so when users bust their v4l2loopback devices with droidcam how are we supported work around that?

Ultimately this issue was why I decided to add that implicit `STREAMON` ioctl in the follow-up patch as the original patch would have broken even more software (e.g. `ffmpeg`).

One automatic workaround OBS could do would be dropping/recreating the affected virtual device (if the module supports this). But this option is also currently only available in the development versions (not released) AFAIR (`v4l2loopback-ctl add/delete /dev/videoX` / ioctl `V4L2LOOPBACK_CTL_ADD` + `V4L2LOOPBACK_CTL_REMOVE`) …

In the end what OBS can do is detect the missing capabilities and either try to perform the above ioctl calls or instruct the user what action they might take.

Not the best situation after all and my impression is that upstream isn't the most quick to respond …

**Comment by kkartaltepe** [comment]
> Not the best situation after all and my impression is that upstream isn't the most quick to respond …

Well if no one in this thread raises the issue with upstream I suspect they will never receive a divination from god that their users are suffering. So back to where we started the best bet is for you people reading this thread to kindly has their package maintainers to patch their packages and to ask for v4l2loopback upstream to respond to this issue affecting them.

**Comment by zDEFz** [comment]
> > Not the best situation after all and my impression is that upstream isn't the most quick to respond …
> 
> Well if no one in this thread raises the issue with upstream I suspect they will never receive a divination from god that their users are suffering. So back to where we started the best bet is for you people reading this thread to kindly ask their package maintainers to patch their packages and to ask for v4l2loopback upstream to respond to this issue affecting them.

Do you want us to post at https://aur.archlinux.org/packages/obs-studio-git or at whatever is here ? https://aur.archlinux.org/packages?O=0&K=v4l2loopback

**Comment by kkartaltepe** [comment]
> Do you want us to post at https://aur.archlinux.org/packages/obs-studio-git or at whatever is here ? https://aur.archlinux.org/packages?O=0&K=v4l2loopback

As mentioned you should report this issue to your v4l2loopback provider. Which v4l2loopback package you are using on arch and how you should report bugs to arch maintainers is beyond what I know, but im sure someone in the arch community can point you in the right direction.

**Comment by zDEFz** [comment]
I did my job :)

**Comment by damiante** [comment]
just popping in to say that I had the same issue (ie virtual camera works exactly once) and the fix worked for me (ie closing OBS, executing `sudo modprobe -r v4l2loopback` in terminal, then restarting OBS).

**Comment by weismannweb** [comment]
@bmn001 this has been an issue of mine for like 2 weeks, I am so grateful to you for that modprobe command it fixed the issue perfectly, i would love to see a proper fix in obs but for now i have your command, thanks so much!

**Comment by zDEFz** [comment]
> @bmn001 this has been an issue of mine for like 2 weeks, I am so grateful to you for that modprobe command it fixed the issue perfectly, i would love to see a proper fix in obs but for now i have your command, thanks so much!

The recommendation was originally by a tip I shared. 
It is not only the virtual camera that does not start/stop properly, but also the camera which is blank if I dont unload / reload uvcvideo!

This is what I call instead of obs:

`
function obs()
{
#This function is intended to prevent blank cameras in OBS upon OBS restart / exit and unmute/remute the mic
#1. load/refresh uvcvideo before starting obs
if lsmod | grep -q 'uvcvideo'; then sudo rmmod uvcvideo; fi 
sudo modprobe uvcvideo
pulseaudio-ctl mute-input #acts as a toogle
#2. since no environment with a keyring to prompt for allowing virtual webcams prior is a must
sudo modprobe v4l2loopback video_nr=10 card_label='OBS Virtual Camera'
	sleep 1
sh -c "obs || sleep 3; sudo rmod uvcvideo; pulseaudio-ctl mute-input"
}
`

**Comment by stephematician** [comment]
The fix that was added, fca624ea4f, is _very non-compliant_ with the V4L2 UAPI. The STREAMON that OBS uses _will_ (and should) fail; see the expected return codes from the V4L2 UAPI [7.55. ioctl VIDIOC_STREAMON, VIDIOC_STREAMOFF](https://www.kernel.org/doc/html/v6.13/userspace-api/media/v4l/vidioc-streamon.html):

> EINVAL
>    The buffer type is not supported, or no buffers have been allocated (memory mapping) or enqueued (output) yet.

I am in the process of hardening v4l2-loopback's compliance with the V4L2 UAPI - and this _will_ return the EINVAL - breaking this fix. Currently, the PR is in draft: https://github.com/umlaeute/v4l2loopback/pull/611 - so there is plenty of time to consider a new fix, as it takes time for everything to downstream.

**Comment by kkartaltepe** [comment]
PRs are welcome, but you should probably make v4l2loopback report its version via the v4l2 api since modinfo is not going to be available on all platforms or even correct since users can update the on disk modules without restarting or reloading.

**Comment by stephematician** [comment]
> PRs are welcome, but you should probably make v4l2loopback report its version via the v4l2 api since modinfo is not going to be available on all platforms or even correct since users can update the on disk modules without restarting or reloading.

I'm not aware of a mechanism within the V4L2 API for reporting module version information - and I'm not an OBS expert so I can't say much about which platforms you support that do have __v4l2loopback__ but _don't_ have __modinfo__ (or __kmod__). The best I can think of is a kludge via [VIDIOC_G_CTRL](https://www.kernel.org/doc/html/v6.13/userspace-api/media/v4l/vidioc-g-ctrl.html) with one control for each of the version numbers; but that's not what controls are meant for.

I'll put up a PR and then the discussion can be moved there.

**Comment by kkartaltepe** [comment]
We only have two supported platforms, flatpak and ubuntu. You can take your guess at which one i'm referring to. Though as a kernel module developer id assume you were aware of the issues folks run into with managing installed and running kernel versions and modules and how tricky it can be.

Last I used v4l2-info it provided a very convenient location for drivers to report their version information. Perhaps you can use that instead of a kludge.

```
v4l2-ctl --info -d /dev/video0
Driver Info:
	Driver name      : uvcvideo
	Card type        : HD Pro Webcam C920
	Bus info         : usb-0000:c4:00.3-2.2
	Driver version   : 6.12.4
```

**Comment by stephematician** [comment]
> Though as a kernel module developer...

I wish I was! Thanks for the help. I had another thought about using sysfs to report the version, that way the user doesn't need __v4l-utils__, but perhaps your approach is easier to implement. I'll get back to this once the __v4l2-loopback__ PR is close.

**Comment by cmosetick** [comment]
This issue has returned.
Can we get it re-opened @kkartaltepe @stephematician ?

**Flatpak OBS version 31.1.2**

I've found this on 2 different machines with different Linux Kernel versions:
5.15  
6.8

Only way to resolve without a full reboot is:

`sudo modprobe -r v4l2loopback`

# Analysis

On system boot you get:
```
ls -la /dev | grep -i video  
crw-rw----   1 root video    29,     0 Aug  7 14:15 fb0
crw-rw----+  1 root video   234,     0 Aug  7 14:15 media0
crw-rw----+  1 root video   234,     1 Aug  7 14:15 media1
crw-rw----+  1 root video    81,     0 Aug  7 14:15 video0
crw-rw----+  1 root video    81,     1 Aug  7 14:15 video1
crw-rw----+  1 root video    81,     2 Aug  7 14:15 video2
crw-rw----+  1 root video    81,     3 Aug  7 14:15 video3
```

Launch the virtual camera in v31.1.2 and then you get a `/dev/video4`

```
ls -la /dev | grep -i video 
crw-rw----   1 root video    29,     0 Aug  7 14:15 fb0
crw-rw----+  1 root video   234,     0 Aug  7 14:15 media0
crw-rw----+  1 root video   234,     1 Aug  7 14:15 media1
crw-rw----+  1 root video    81,     0 Aug  7 14:15 video0
crw-rw----+  1 root video    81,     1 Aug  7 14:15 video1
crw-rw----+  1 root video    81,     2 Aug  7 14:15 video2
crw-rw----+  1 root video    81,     3 Aug  7 14:15 video3
crw-rw----+  1 root video    81,     4 Aug  7 14:42 video4
```

If you stop the virtual camera now, the next time that you launch it via OBS, you will get errors:

```
info: NV12 texture support enabled
info: P010 texture support not available
info: Attempting to reset output capability of '/dev/video0'
info: Attempting to reset output capability of '/dev/video1'
info: Attempting to reset output capability of '/dev/video2'
info: Attempting to reset output capability of '/dev/video3'
info: Attempting to reset output capability of '/dev/video4'
warning: Failed to start virtual camera
```

# Fix without a full reboot

`sudo modprobe -r v4l2loopback`

This command will remove `/dev/video4`

And now the virtual camera will start again without a full reboot. 

Note that simply doing a `sudo rm /dev/video4` will not resolve.

The module reload seems required and should be resolved.

**Comment by stephematician** [comment]
@cmosetick I would try the forums for support (I'm checking there as often as I can) as this may not be an issue with OBS and there's a lot of detail about your system and use case is missing (NOTE: the kernel version is almost irrelevant, because v4l2loopback is an out-of-tree module).

Also NOTE: The command `rm /dev/video4` is _definitely_ not a solution, and I would advise against anyone using an `rm` command on any contents of the `/dev` folder unless they have a _really_ specific reason for doing so.

**Comment by jbrains** [comment]
FWIW, I experienced this problem before, but I don't experience it now. Here is what I can repeatably do:

1. remove module `v4l2loopback`.
2. start OBS Studio, Flatpak, 31.1.2
3. press "Start Virtual Camera". This loads `v4l2loopback` and I see a new video device in `/dev`.
4. press "Stop Virtual Camera". The Virtual Camera video device remains listed in `/dev`.
5. press "Start Virtual Camera". Nothing goes wrong and the Virtual Camera device works in Zoom. There are no OBS Studio log messages that indicate a failure.

My kernel is 6.12.10.

I hope this information helps somehow. I second @stephematician 's suggestion to move this into the forum. Post a link here to the thread when you do.

**Comment by Atom02** [comment]
ye it is comming back and the only "work around" is to run sudo modprobe -r v4l2loopback everytime i want to start virtual camera

**Comment by mrbarletta** [comment]
Its happening in Ubuntu 24.04 with version 32.1.0

```
info: Kernel Version: Linux 6.17.0-19-generic
info: Distribution: "KDE neon" "24.04"
info: Desktop Environment: KDE (KDE)
info: Session Type: wayland
info: Current Date/Time: 2026-04-01
```

**Comment by zDEFz** [comment]
> Its happening in Ubuntu 24.04 with version 32.1.0
> 
> ```
> info: Kernel Version: Linux 6.17.0-19-generic
> info: Distribution: "KDE neon" "24.04"
> info: Desktop Environment: KDE (KDE)
> info: Session Type: wayland
> info: Current Date/Time: 2026-04-01
> ```

Yea, you can absolutely fix it manually.
Not sure why this issue is closed

**Comment by stephematician** [comment]
@zDEFz it's closed because it was closed - there's a new issue because it's failing for a different reason now, see: https://github.com/obsproject/obs-studio/issues/12506. Generally speaking it's because some distributions are still using older _v4l2loopback_. @mrbarletta - if you can install a newer version of _v4l2loopback_, that would likely resolve the issue.
